lib/capistrano/asg.rb

```diff
@@ -38,22 +39,43 @@ def autoscale(groupname, *args)
   (regions[region] ||= []) << groupname
   set :regions, regions
 
+  # Create an array of role names to be distributed across the ASG
+  partial_queue = []
+  if args.key?(:partial_roles)
+    args[:partial_roles].each do |partial|
+      instances = partial.key?(:instances) ? partial[:instances] : 1
+      instances.times { partial_queue << partial[:name].to_s }
+    end
+    args.delete(:partial_roles)
+  end
+
   asg_instances.each do |asg_instance|
     if asg_instance.health_status != 'Healthy'
       puts "Autoscaling: Skipping unhealthy instance #{asg_instance.id}"
     else
-      ec2_instance = ec2_resource.instance(asg_instance.id)
-      hostname = ec2_instance.private_ip_address
-      puts "Autoscaling: Adding server #{hostname}"
-      server(hostname, *args)
+      with_retry do
+        ec2_instance = ec2_resource.instance(asg_instance.id)
+        hostname = ec2_instance.private_ip_address
+        puts "Autoscaling: Adding server #{hostname}"
+        # create a complete temp copy of the array contents instead of just copying the references
+        host_args = Marshal.load(Marshal.dump(args))
```

Comment:
Doing things like this is considered bad practice and insecure. It's also unnecessary, we should just pass along the parameters that are meant for the method.

Reply:
What I ran into when trying to make a copy of the `args` array is that it was copying references, so when I updated the temp `host_args` it also updated the values in the original array.  Looks like there are some gems out there to accomplish this - https://stackoverflow.com/questions/8206523/how-to-create-a-deep-copy-of-an-object-in-ruby

What do you suggest?